src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs

```diff
@@ -0,0 +1,214 @@
+                        {
+                            numberOfCharsToIgnore++;
+                        }
+                        else if (charToValidate == EncodingPad)
+                        {
+                            paddingCount++;
+                        }
+                    }
+
+                    length -= numberOfCharsToIgnore;
+                }
+            }
+            else if (length >= 4)
+            {
+                if (base64Text[length - 1] == EncodingPad)
+                {
+                    paddingCount++;
+                }
+                if (base64Text[length - 2] == EncodingPad)
+                {
+                    paddingCount++;
+                }
+            }
+
+            if (length % 4 != 0)
+            {
+                decodedLength = 0;
+                return false;
+            }
+
+            // Remove padding to get exact length
+            decodedLength = (length / 4 * 3) - paddingCount;
+            return true;
+        }
+
+        /// <summary>
+        /// Validates the span of UTF-8 encoded text represented as base64 into binary data.
+        /// </summary>
+        /// <param name="base64TextUtf8">The input span which contains UTF-8 encoded text in base64 that needs to be validated.</param>
+        /// <returns>true if <paramref name="base64TextUtf8"/> is decodable; otherwise, false.</returns>
+        public static bool IsValid(ReadOnlySpan<byte> base64TextUtf8) => IsValid(base64TextUtf8, out int _);
+
+        /// <summary>
+        /// Validates the span of UTF-8 encoded text represented as base64 into binary data.
+        /// </summary>
+        /// <param name="base64TextUtf8">The input span which contains UTF-8 encoded text in base64 that needs to be validated.</param>
+        /// <param name="decodedLength">The maximum length (in bytes) if you were to decode the base 64 encoded text <paramref name="base64TextUtf8"/> within a byte span.</param>
+        /// <returns>true if <paramref name="base64TextUtf8"/> is decodable; otherwise, false.</returns>
+        public static unsafe bool IsValid(ReadOnlySpan<byte> base64TextUtf8, out int decodedLength)
+        {
+            if (base64TextUtf8.IsEmpty)
+            {
+                decodedLength = 0;
+                return true;
+            }
+
+            // Check for invalid chars
+            int indexOfFirstNonBase64 = base64TextUtf8.IndexOfAnyExcept(ValidBase64BytesSortedAsc);
+            if (indexOfFirstNonBase64 >= 0)
+            {
+                decodedLength = 0;
+                return false;
+            }
+
+            int length = base64TextUtf8.Length;
+            int paddingCount = 0;
+
+            // Check if there are chars that need to be ignored while determining the length
+            if (base64TextUtf8.IndexOfAny(BytesToIgnore) > -1)
+            {
+                fixed (byte* srcBytes = &MemoryMarshal.GetReference(base64TextUtf8))
```

Comment:
There should be no need to switch to unsafe code here, a simple `foreach` over the span will be just as fast.